src/modules/commander/state_machine_helper.cpp

```diff
@@ -781,30 +808,25 @@ bool check_invalid_pos_nav_state(vehicle_status_s &status, bool old_failsafe, or
 
 		} else {
 			// go into a descent that does not require stick control
-			if (status_flags.condition_local_position_valid) {
+			if (status.vehicle_type == vehicle_status_s::VEHICLE_TYPE_ROTARY_WING &&
+			    status_flags.condition_local_position_valid) {
 				status.nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_LAND;
 
-			} else if (status_flags.condition_local_altitude_valid) {
-				if (status.vehicle_type == vehicle_status_s::VEHICLE_TYPE_ROTARY_WING) {
+			} else  if (status_flags.condition_local_altitude_valid) {
+				// Switch to Descend state if MC or if FW and loitering time has passed
+				if (status.vehicle_type == vehicle_status_s::VEHICLE_TYPE_ROTARY_WING ||
+				    hrt_elapsed_time(&status.failsafe_timestamp) > (gps_fail_openloop_loiter_time_s * 1_s)) {
+
 					status.nav_state = vehicle_status_s::NAVIGATION_STATE_DESCEND;
 
 				} else {
-					// TODO: FW position controller doesn't run without condition_global_position_valid
-					status.nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_LANDGPSFAIL;
```

Comment:
@sfuhrer If you search the code base you will find that we also switch into this nav state at other locations. Do we need to handle those as well?

Reply:
Right, it's currently also used in set_link_loss_nav_state(), set_offboard_loss_nav_state(), set_offboard_loss_rc_nav_state(). I don't see a reason to not use the same updated logic around NAVIGATION_STATE_FIXED_BANK_LOITER there as well, do you?